bittide-instances/src/Bittide/Instances/Tests/FullMeshHwCc.hs

```diff
@@ -327,57 +327,65 @@ fullMeshHwCcTest refClkDiff sysClkDiff syncIn rxns rxps miso =
   (sysClk, sysLock0) = clockWizardDifferential (SSymbol @"SysClk") sysClkDiff noReset
   sysLock1 = xpmCdcSingle sysClk sysClk sysLock0 -- improvised reset syncer
   sysRst = unsafeFromActiveLow sysLock1
-
-  syncIn1 = (startTest .&&.)
-    $ unsafeToActiveLow
+    `orReset` unsafeFromActiveLow startTest  -- <--------------------------|
+  --                                                                       |
+  -- 'syncOutGenerator' is used to drive the 'SYNC_OUT' signal, which      |
+  -- is only connected for the last node in the network and wired back     |
+  -- to 'SYNC_IN' of all nodes from there.                                 |
+  --                                                                       |
+  -- Note that all nodes are in reset before their local 'startTest' VIO   |
+  -- signal gets asserted, as 'startTest' is directly driving 'sysRst'. ----
+  -- Thus, for the other nodes to capture the 'SYNC_OUT' signal correctly,
+  -- the node receiving the `startTest` rising edge last must be the one
+  -- with it's 'SYNC_OUT' physically connected to the 'SYNC_IN' of all
+  -- nodes in the network. This assumption is tested by
+  -- 'Bittide.Instances.Tests.SyncInSyncOut'.
+  syncOut = syncOutGenerator sysClk startTest
```

Comment:
> (See previous comment)
>>  We don't need to introduce a delay. I solved the problem more elegantly now.

How did you fix the potential glitches on `syncOut`? As a rule of thumb, any combinatorial logic connected to a pin can close logic. (This isn't entirely true, but it's so hard to predict that it is good to keep this as a rule.)

Reply:
[That line](https://github.com/bittide/bittide-hardware/blob/6335a6ebdf07b4aacf24599248cb422e4302bef8/bittide-instances/src/Bittide/Instances/Tests/FullMeshHwCc/IlaPlot.hs#L256C4-L256C4) should be fixing it, shouldn't it?

Reply:
That could in theory create even more glitches! The problem here is that signals are allowed to "settle" between their hold and setup window. That is, the window marked with `XXXX` in this diagram:

![setup and hold timing checks](https://github.com/bittide/bittide-hardware/assets/5658854/542e0bab-7fb5-42d1-a412-e779542ec0c5)

So _in theory_ the VIO could output some complicated logic, making it move `startTest` to various values before settling on its final value. The same goes for the output of your mealy machine: its implemented using a bunch of muxes whose outputs might be mapped two different path lengths causing race conditions / glitches. In general we don't care, because glitches are okay within that time window / in the synchronous abstraction. However, when connected to a pin this makes the other side potentially see all the in-between states!

Check out [Glitches in combinational circuits ](https://vlsiuniverse.blogspot.com/2017/10/glitches-in-combinational-circuits.html) for a better explanation.

Because even trivial circuits can trigger glitches, we generally just use a registered output (`delay`) unless there is a very good reason to believe the circuit doesn't produce glitches + an extra register would mess up the design.

Reply:
As an example of how this can go wrong, this function:

https://github.com/bittide/bittide-hardware/blob/1338fd3d95a297a36a950dba9f7ed97f6eabe0c4/bittide/src/Bittide/ClockControl.hs#L187-L188

Used to _not_ have a `dflipflop`, which made the clock boards see very short FINC/FDEC pulses. This messed up clock control quite a bit..

Reply:
Alright, gotcha. So this is basically not releated to `syncOutGenerator` at all, because `SYNC_OUT` is really directly driving the output pin here. I always thought this is handled somewhere later. So [in fact it wasn't even handled correctly](https://github.com/bittide/bittide-hardware/blob/f23b4e80838d949ba38a8828e8b1539d129c5324/bittide-instances/src/Bittide/Instances/Tests/FullMeshHwCc.hs#L304) before merging #399 already.

I alwasy misunderstood this in the sense that `syncOutGenerator` would require even another additional latch beside that one for the output  pin.

Reply:
Right, `syncOutGenerator` is fine on its own. Though you could make the argument that because its, well, the SYNC_OUT generator, it should just produce glitchless signals.